src/igraph_eulerian.c

```diff
@@ -0,0 +1,660 @@
+/* -*- mode: C -*-  */
+/* vim:set ts=4 sw=4 sts=4 et: */
+/*
+   IGraph library.
+   Copyright (C) 2005-2012  Gabor Csardi <csardi.gabor@gmail.com>
+   334 Harvard street, Cambridge, MA 02139 USA
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+   You should have received a copy of the GNU General Public License
+   along with this program; if not, write to the Free Software
+   Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+   02110-1301 USA
+*/
+
+#include "igraph_eulerian.h"
+#include "igraph_adjlist.h"
+#include "igraph_interface.h"
+#include "igraph_components.h"
+#include "igraph_types_internal.h"
+
+/**
+ * \section about_eulerian
+ *
+ * <para>These functions calculate whether an Eulerian path or cycle exists
+ * and if so, can find them.</para>
+ */
+
+
+/* solution adapted from https://www.geeksforgeeks.org/eulerian-path-and-circuit/
+The function returns one of the following values
+has_path is set to 1 if a path exists, 0 otherwise
+has_cycle is set to 1 if a cycle exists, 0 otherwise
+*/
+static int igraph_i_is_eulerian_undirected(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle, igraph_integer_t *start_of_path) {
+    igraph_integer_t odd;
+    igraph_vector_t degree, csize, check_for_self_loops, cluster_member;
+    igraph_inclist_t il;
+    long int i, j, vsize;
+    long int cluster_count;
+
+    if (igraph_ecount(graph) == 0 || igraph_vcount(graph) <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return  IGRAPH_SUCCESS;
+    }
```

Comment:
Whenever the function returns with `has_path` being true, and the graph has at least one vertex, it should not leave `start_of_path` un-set. The graph may have one vertex with many self-loops, so `start_of_path` may end up being used.

Reply:
Just fixed this.